atc/engine/builder/builder.go

```diff
@@ -33,7 +34,7 @@ type StepFactory interface {
 type DelegateFactory interface {
 	GetDelegate(db.Build, atc.PlanID, vars.CredVarsTracker) exec.GetDelegate
 	PutDelegate(db.Build, atc.PlanID, vars.CredVarsTracker) exec.PutDelegate
-	TaskDelegate(db.Build, atc.PlanID, vars.CredVarsTracker) exec.TaskDelegate
+	TaskDelegate(db.Build, atc.PlanID, vars.CredVarsTracker, policy.Checker) exec.TaskDelegate
```

Comment:
@pivotal-jwinters Can you please be more specific? If you want to apply policy check to all steps?

Reply:
Didn't you outline some of these things in your RFC? Like being able to restrict where an image gets pulled from? Since an image can come from get step we would need to enforce these policies at that level.

I'm also curious why the policy.Checker gets passed through the delegate. Could we not pass it directly to the step itself?

Reply:
@pivotal-jwinters  I think I explained in the RFC: https://github.com/concourse/rfcs/pull/41/files#diff-920d7a06e64dac5ea2673094189fa570R38. With `set-pipeline` API, everything about a pipeline, expect task files, are available to check. So docker-images used for `get` and `put` can be checked at `set-pipeline`. Because task files are only fetched at runtime, to check images used in `task` step, I have to add a check point in `task` step.

I received a comment for a previous PR: https://github.com/concourse/concourse/pull/4708#pullrequestreview-311954465 from you that db related operation should be delegate, so I think policy related operations should be similar. Is there a guide for what should be in delegate and what should be in step itself?

Reply:
Oh so you want to send the entire set-pipeline payload to be verified by the policy server? That means the policy server will need to know how to interpret a pipeline yaml? Wouldn't it be a lot easier to make these policy decisions at the get/put level instead?

There's no guide, but the way i look at it is - anything that controls the flow of logic should be within the step itself, and anything that deals with side effects (stdout, saving versions) should be in the delegate. So the delegate ends up dealing with a lot of db operations.

Overall the exec steps are very much a work in progress. In the past we've found that making large changes to this part of the code is very difficult (in part) because the steps are all very different. Over time we've been trying to align these steps to be more consistent, so these one off changes tend to raise red flags for me.

Reply:
> Oh so you want to send the entire set-pipeline payload to be verified by the policy server? 

Yes, all API calls' payload will be sent to the policy server, unless filtered (there are some filters to prevent from some APIs from going through the policy server.)

> That means the policy server will need to know how to interpret a pipeline yaml? Wouldn't it be a lot easier to make these policy decisions at the get/put level instead?

No, that's not true. To the policy server, every input is a JSON. So checking an entire pipeline is almost same easy as checking a task config.

But after thinking over, your comment is valid from the other perspective. Every get and put should also go through policy check, so that policy change may immediately take effective.  Otherwise if we only do policy check for `SetPipeline`, when policy is changed, say some image is no longer allowed to use, the policy change will not be able to apply to existing pipelines.

For that, I'll update the code as well as the RFC.

> There's no guide, but the way i look at it is - anything that controls the flow of logic should be within the step itself, and anything that deals with side effects (stdout, saving versions) should be in the delegate. So the delegate ends up dealing with a lot of db operations.
> 
> Overall the exec steps are very much a work in progress. In the past we've found that making large changes to this part of the code is very difficult (in part) because the steps are all very different. Over time we've been trying to align these steps to be more consistent, so these one off changes tend to raise red flags for me.

OK, I'll move policy check to step.

Reply:
@pivotal-jwinters I have pushed a new commit that moved `polichChecker` from delegate to step, and added policy check to `get` and `put`. I renamed the action name "RunTask" to "UsingImage", if you have a better name, feel free to advice. Please take a look again and see if you are ok with the new change.